yarn-project/simulator/src/avm/journal/trace.ts

```diff
@@ -12,38 +31,137 @@ import {
   type TracedPublicStorageWrite,
 } from './trace_types.js';
 
+export class TraceArray<T> extends Array<T> {
+  /** items explicitly pushed to this instance, not including parent's, not including items "merged" in */
+  public thisLength: number = 0;
+
+  constructor(
+    private name: string,
+    private maxThisLen: number,
+    private maxTotalLen: number,
```

Comment:
I think you can make these generic with `MaxLen extends number` etc and have them baked into the type.

Reply:
Otherwise tuples can be used: https://stackoverflow.com/questions/41139763/how-to-declare-a-fixed-length-array-in-typescript